tasmota/xsns_96_bl6523.ino

```diff
@@ -0,0 +1,353 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef USE_BL6523
+/*********************************************************************************************\
+ * Chinese BL6523  based Watt hour meter
+ *
+ * This meter provides accurate Voltage, Frequency, Ampere, Wattage, Power Factor, KWh 
+ * To use Tasmota the user needs to add an ESP8266 or ESP32
+ * Three lines need to be connected via 1KOhh resistors to ESP from the main board(RX,TX GND)
+ * 
+ * Connection Eg (ESP8266) - Non - Isolated:
+ *    BL6523 RX ->1KOhm-> ESP IO4(D2) (Should be Input Capable)
+ *    BL6523 TX ->1KOhm->  ESP IO5(D1) (Should be Input Capable)
+ *    BL6523 GND -> ESP GND
+ * 
+ * Connection Eg (ESP32) - Non - Isolated:
+ *    BL6523 RX ->1KOhm-> ESP IO4 (Should be Input Capable)
+ *    BL6523 TX ->1KOhm-> ESP IO5 (Should be Input Capable)
+ *    BL6523 GND -> ESP GND
+ * 
+ * To build add the below to user_config_override.h
+ * #define USE_BL6523       // Add support for Chinese BL6523 based Watt hour meter (+1k code)¸
+ * 
+ * After Installation use the below template sample:
+ * {"NAME":"BL6523 Smart Meter","GPIO":[0,0,0,0,7488,7520,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":18}
+\*********************************************************************************************/
+
+#define XSNS_96 96
+
+#include <TasmotaSerial.h>
+
+#define BL6523_RX_DATASET_SIZE 2
+#define BL6523_TX_DATASET_SIZE 4
+
+#define BL6523_BAUD 4800
+#define BL6523_REG_AMPS 0x05
+#define BL6523_REG_VOLTS 0x07
+#define BL6523_REG_FREQ  0x09
+#define BL6523_REG_WATTS 0x0A
+#define BL6523_REG_POWF  0x08
+#define BL6523_REG_WATTHR 0x0C
+
+/* No idea how to derive human readable units from the byte stream.
+For now  dividing the 24-bit values with below constants seems to yield something sane
+that matches whatever displayed in the screen of my 240v model. Probably it would be possible
+to extract these values from the write register commands (0xCA).*/ 
+#define BL6523_DIV_AMPS 297899.4f
+#define BL6523_DIV_VOLTS 13304.0f
+#define BL6523_DIV_FREQ  3907.0f
+#define BL6523_DIV_WATTS 707.0f
+#define BL6523_DIV_WATTHR 674.0f
+
+TasmotaSerial *Bl6523RxSerial;
+TasmotaSerial *Bl6523TxSerial;
+
+struct BL6523
+{
+  uint32_t amps = 0;
+  uint32_t volts = 0;
+  uint32_t freq = 0;
+  uint32_t watts = 0;
+  uint32_t powf = 0;
+  uint32_t watthr = 0;
+
+  uint8_t type = 1;
+  uint8_t valid = 0;
+  uint8_t got_data_stone = 0;
+  bool discovery_triggered = false;
+} Bl6523;
```

Comment:
INstead of statically allocate memory, please make that a pointer to the structure and allocate memory only if the driver is in use
That will save RAM for when the driver is compiled in a large build but not used.